### ISSUE TITLE
TCP Connection reset after connection was idle for more that 15 minutes

### ISSUE THREAD

**Issue by tanapoln** [cloudnativelabs/kube-router#521]
My applications communicates via gRPC and the connections was not active all time. When a specific connection was idle for more than 15 minutes, it would get RST packet after try to write to that connection.

A tcpdump on client-side (a sender) shows that RST packet comes from remote IP but tcpdump on server-side (a receiver) show nothing about RST packet. I also notice that TTL of RST packet is not decrease (default TTL=64 in my server).

I suspect that it was not sent from remote but in a local itself. You can check with this tcpdump capture: [tcpdump.zip](https://github.com/cloudnativelabs/kube-router/files/2320742/tcpdump.zip)

I also wonder why there are retransmission and duplicate packets in tcpdump result but leave it later about that.

The problem I found can reproduce reliably, you can use attached Kubernetes YAML to test and see result: [grpc-test.yml.zip](https://github.com/cloudnativelabs/kube-router/files/2320741/grpc-test.yml.zip). I also tested using default `kubenet` and problem was not found.

The code of grpc-test do simply task: client send RPC to server every specific interval which can be configure via environment variable `INTERVAL_MS`. The connection will establish at the beginning when client was started.

### Environments:
Kubernetes: 1.9.3
Docker: 17.3.2
Linux Kernel: 4.4.121-k8s (Debian jessie)
Kube-router: 0.1.0

My cluster was provisioned using Kops with this command:
```
export NAME=testcluster.k8s.local
export KOPS_STATE_STORE=s3://my-testcluster-kops-xxx
kops create cluster --dns private --api-loadbalancer-type public --kubernetes-version=1.9.3 --zones ap-southeast-1b --node-count 1 --node-size t2.medium --master-size t2.medium --networking kube-router -t public ${NAME}
```

**Comment by tanapoln** [comment]
Investigated and found a problem.

The default value of IPVS Connection timeout is 900 seconds, after increase a timeout that (via `ipvsadm --set`) a problem is gone.

I think this config value must be implemented inside kube-route due to a manual config is not persist across reboot.

**Comment by znoodl** [comment]
@murali-reddy 
I found another way to fix this, https://success.docker.com/article/ipvs-connection-timeout-issue
> tcp_keepalive_time must less than ipvs(900 secs), and client socket must set SO_KEEPALIVE.

But it's not works for me, do you have some plans to support config in kube-router?
or I have to set it by myself.

**Comment by roffe** [comment]
@znoodl Good find, i think we could implement a lower "net.ipv4.tcp_keepalive_time" without side effects here https://github.com/cloudnativelabs/kube-router/blob/master/pkg/controllers/proxy/network_services_controller.go#L278-L319

@murali-reddy @andrewsykim any thoughts on lowering the value. On my kops nodes the default is `7200`

**Comment by znoodl** [comment]
> @znoodl Good find, i think we could implement a lower `net.ipv4.tcp_keepalive_time=600` without side effects here https://github.com/cloudnativelabs/kube-router/blob/master/pkg/controllers/proxy/network_services_controller.go#L278-L319
> 
> @murali-reddy @andrewsykim any thoughts on lowering the value. On my kops nodes the default is `7200`

I think you misunderstand.
For me, `net.ipv4.tcp_keepalive_time=600` not works, it should be test more before merge to kube-router.
And supporting ipvs timeout (command like `ipvsadm --set 7500 120 300`) config in kube-router is better.

**Comment by roffe** [comment]
@znoodl for tcp_keepalive_time to work your client must utilise it, the change make sense IMHO even if it didn't work for you

**Comment by roffe** [comment]
I think for gRPC you should configure https://github.com/grpc/grpc-go/blob/master/keepalive/keepalive.go

**Comment by znoodl** [comment]
Yes, let program to do this is the best way, but it should have default config to avoid `Connection reset ` happens too frequently ( default IPVS timeout is 900s).

**Comment by roffe** [comment]
@znoodl i think you misunderstand. tcp_keepalive_time is not doing anything unless the client connecting has set SO_KEEPALIVE on the connection, in gRPC's case, I can't find that it does it but has it's own ping/keepalive mechanism on the wire.

**Comment by znoodl** [comment]
@roffe Thank you, I figure it out. Now I set tcp_keepalive in my applications, it's working well.
